File: platforms/ios/runtime/Sources/NativeblocksRuntime/frame/FrameStateBridge.swift
import Foundation
import NativeblocksRuntimeFFI

internal protocol FrameStateBridge {
    func observeFrame(
        route: String,
        args: [String: String],
        stateKey: String?,
        onFull: @escaping (FrameFull) -> Void,
        onDiff: @escaping (FrameDiff) -> Void
    ) async

    func updateVariable(key: String, value: String)


    func logAction(event: ActionLogEvent)

    func logBlock(event: BlockLogEvent)

    func releaseFrame()
}

internal final class FrameStateBridgeImpl: FrameStateBridge {

    private let frameStateManager: FrameStateManager

    init(frameStateManager: FrameStateManager) {
        self.frameStateManager = frameStateManager
    }

    func observeFrame(
        route: String,
        args: [String: String],
        stateKey: String?,
        onFull: @escaping (FrameFull) -> Void,
        onDiff: @escaping (FrameDiff) -> Void
    ) async {
        await frameStateManager.setupFrame(
            route: route,
            args: args,
            stateKey: stateKey,
            observer: FrameStateObserverAdapter(onFull: onFull, onDiff: onDiff)
        )
    }

    func updateVariable(key: String, value: String) {
        frameStateManager.updateVariable(key: key, value: value)
    }


    func logAction(event: ActionLogEvent) {
        frameStateManager.logAction(event: event)
    }

    func logBlock(event: BlockLogEvent) {
        frameStateManager.logBlock(event: event)
    }

    func releaseFrame() {
        frameStateManager.release()
    }
}

private final class FrameStateObserverAdapter: FrameStateObserver, @unchecked Sendable {

    private let onFull: (FrameFull) -> Void
    private let onDiff: (FrameDiff) -> Void

    init(onFull: @escaping (FrameFull) -> Void, onDiff: @escaping (FrameDiff) -> Void) {
        self.onFull = onFull
        self.onDiff = onDiff
    }

    func onFrameChange(change: FrameChangeType) {
        switch change {
        case .full(let frame): onFull(frame)
        case .diff(let frame): onDiff(frame)
        }
    }
}

File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/frame/FrameStateBridge.kt
package io.nativeblocks.runtime.frame

import io.nativeblocks.runtime.ffi.ActionLogEvent
import io.nativeblocks.runtime.ffi.FrameChangeType
import io.nativeblocks.runtime.ffi.FrameDiff
import io.nativeblocks.runtime.ffi.FrameFull
import io.nativeblocks.runtime.ffi.FrameStateManager
import io.nativeblocks.runtime.ffi.FrameStateObserver

internal interface FrameStateBridge {
    suspend fun observeFrame(
        route: String,
        args: Map<String, String>,
        stateKey: String?,
        onFull: (FrameFull) -> Unit,
        onDiff: (FrameDiff) -> Unit,
    )

    fun updateVariable(key: String, value: String)

    fun updateBlockData(blockKey: String, dataKey: String, value: String)

    fun updateBlockProperty(
        blockKey: String,
        propertyKey: String,
        valueMobile: String,
        valueTablet: String,
        valueDesktop: String
    )

    fun logAction(event: ActionLogEvent)

    fun releaseFrame()
}

internal class FrameStateBridgeImpl(
    private val frameStateManager: FrameStateManager,
) : FrameStateBridge {

    override suspend fun observeFrame(
        route: String,
        args: Map<String, String>,
        stateKey: String?,
        onFull: (FrameFull) -> Unit,
        onDiff: (FrameDiff) -> Unit,
    ) {
        frameStateManager.setupFrame(route, args, stateKey, object : FrameStateObserver {
            override fun onFrameChange(change: FrameChangeType) {
                when (change) {
                    is FrameChangeType.Full -> onFull(change.frame)
                    is FrameChangeType.Diff -> onDiff(change.frame)
                }
            }
        })
    }

    override fun updateVariable(key: String, value: String) {
        frameStateManager.updateVariable(key, value)
    }

    override fun updateBlockData(blockKey: String, dataKey: String, value: String) {
        frameStateManager.updateBlockData(blockKey, dataKey, value)
    }

    override fun updateBlockProperty(
        blockKey: String,
        propertyKey: String,
        valueMobile: String,
        valueTablet: String,
        valueDesktop: String
    ) {
        frameStateManager.updateBlockProperty(
            blockKey,
            propertyKey,
            valueMobile,
            valueTablet,
            valueDesktop
        )
    }

    override fun logAction(event: ActionLogEvent) {
        frameStateManager.logAction(event)
    }

    override fun releaseFrame() {
        frameStateManager.release()
    }
}
